---
title: "CS510 Project"
author: "Created by Yesha"
output: html_document
---

###Study Aim
The aim of the script below is to run genetic analyses (a mini GWAS) to identify genetic variants that are associated with breast cancer risk. Particularly we are looking at a sample of 3,135 genetic variants among ~100K European individuals from around the world who were enrolled in this study from various study cites. The output of the analyses will be a tab delimited .txt file with test statistics as well as Manhattan and Q-Q plots which will display our results to see if any markers are found to be significantly associated across the genome.

Note: Although, the data is confidential and cannot be shared due to IRB restrictions, it might help to reference that it is a subset of the Breast Cancer Association Consortium detailed here: http://bcac.ccge.medschl.cam.ac.uk/

##Preamble
### Download the necessary packages

```{r, warning = FALSE, echo = T, results = 'hide'}
list.of.packages <- c("ggplot2", "data.table", "statmod", "qqman", "tidyr")
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) install.packages(new.packages)

# Open the packages
library(ggplot2)
library(data.table)
library(statmod)
library(tidyr)
library(qqman)
```
### Read the data
```{r, warning = FALSE}
setwd ("C:\\Users\\yadmin\\Desktop\\Other\\Chapman_Program\\Classes\\CS-510-Computing_for_Scientists\\Project")
```

#####The dosage file
This is where the genetic data is stored per person per genetic variant, currently the rows are the variants and the columns are the samples - we will have to transpose this later to run the statistical analyses. Also note, in this form (given to us by the investigator) none of the sample names have been included in the genetic data files. Dosage values range from 0 1o 1 starting column V3 below. V1 = genetic variant ID and V2 = genetic variant position

```{r, warning = FALSE}
dose <- fread("BCAC_582_imputed_dosages.txt", sep=" ", header=F, fill=TRUE)
dose[1:5,1:9]
```

#####Read in the phenotype & covariate file
"status_new" is the case/control status data, PC1-PC10 represent principal components (further explained below)

```{r, warning = FALSE}
covar <- read.table("BCAC_582_pheno_yp.txt", header=TRUE, fill=T) 
covar [1:5, 3:18] #Note: I've purposely omitted "printing" the ID colummns, col 1 & 2
```

### Data Management
The sample IDs for the genetic/dosage data were provided separately in a .txt file - so we have to merge in this information before we can link the genetic data with any covariate information related to each of the samples. 

```{r}
names <-fread("BCAC_582_sample_order.txt", header=F)
names(dose)<-c("snp","bp","a1","a2",names$V1)
```

The first three columns of the dosage files provide information about the variants (their chromosomal location, and the two alleles realted to the variant), we don't need this at this point, so we will drop the first three columns. Further, removing this informaton below will allow us to smoothly transpose the data next

```{r}
dose2 <-subset(dose, select= -c(bp, a1, a2))
name2 <- dose2$snp
```
In order to analyze the data, we have to convert the rows (variants) into columns, and columns (samples) into rows - this will allows us to analyze each variant one at a time for all samples. To do so, transpose all but the first column (name)

```{r, echo = T, results = 'hide'}
trans <- as.data.frame(t(dose2[,-1]))
colnames(trans) <- name2

setDT(trans, keep.rownames = TRUE)[]  
colnames(trans)[1] <- "BCAC_ID"   # converts the row names which contain the IDs into an actual column in the dataset
```
```{r}
trans[1:5,2:5]   #Note: I've purposely omitted "printing" the ID colummn, col 1
```

Now that all of the data is in the same format (where samples are the rows) we can merge the covariates, phenotype data with the genetic dosage data
```{r, warning = FALSE}
all <- merge(covar, trans, by.x="BCAC_ID", by.y="BCAC_ID", sort = TRUE)
all[1:5,3:15] #Note: I've purposely omitted "printing" the ID colummns, col 1 & 2
all[1:5,17:20] #The dosage values begin starting column 19
```

### Analysis
Prior to running the analysis let's first identify the number of cases (53,772) & controls (76,479); this is a large study with 130,251 samples
```{r, warning = FALSE}
table(all$status_new)
```

In order to identify the association between a given variant / dosage value with case/control status, we will run a logistic regression model (glm, family = logit) and additionally adjust for the country of origin of study which is known to affect differences in associations, and the first four principal components (which are more precise genetic markers used to identify race/ethnicity - rather than using self identified race/ethnicity)

First we fit a null model where dosage is not included and there after each iteration below within the loop will cycle through every variant, with the start point at 19

```{r, warning = FALSE}
reg.null <- glm (status_new ~ as.factor(StudyCountry) + PC1 + PC2 + PC3 + PC4, family=binomial, data=all)
summary(reg.null)
```

As seen in the regression results above, the country of study origin is signfiicantly varied depending on the country in question in realtion to cancer risk and this is due to population stratification issues which is why we are adjusting for study country in our analyses as well as the PC values.

The loop below will take in the dosage value for each person one variant at a time - using this we can then calculate the frequency of the variant (i) in our study here (for comparison of its occurrence in the general population), and conduct the following analyses: the likelihood ratio test, the score test and the wald test - all three of which are necessary test statistics for various components of the variant tests as explained below  

```{r setup, include=TRUE, warning = FALSE}
knitr::opts_chunk$set(cache = TRUE)

start <- 19  #This value can be adjusted for any subset of variants you want to analyze
stop <- dim(all)[2]
dim(all)

for (i in start:stop){
  
  Dosage <- all[i]
  Dosage <- apply(Dosage, 1, as.numeric)
  
# Determine the Frequency of each variant in this dataset
  Frequency <- mean(Dosage)/2
  
# Analyze the data using the Score Test - more appropriate test measure for variants that are rare in our data - this is the proper test statistic to report 
  Score.Z <-glm.scoretest(reg.null,Dosage,dispersion=NULL)
  P.Score <-2*(1-pnorm(abs(Score.Z)))
  
# Analyze the data using the Likelihood Ratio Test for comparison & general use
  reg <- glm(status_new ~ as.factor(StudyCountry) + PC1 + PC2 + PC3 + PC4 + Dosage, family=binomial, data=all)
  
  chi.stat.LRT <- 2*(logLik(reg) - logLik(reg.null))
  P.LRT <- 1-pchisq(chi.stat.LRT, df=1)
  
# If Dosage is missing for all samples for a given variant, then we must indicate the missingness in our results with "NA" for the beta estimate, SE and P.Wald as done below, otherwise report the test statistics as they are being calculated
    if (!"Dosage" %in% rownames(summary(reg)$coefficients)){
    Estimate <- NA
    SE <- NA
    P.Wald <- NA
  }
  
  if ("Dosage" %in% rownames(summary(reg)$coefficients)){
    Estimate <- coef(summary(reg))["Dosage","Estimate"]  #extract the beta estimate value from the regression output
    SE <- coef(summary(reg))["Dosage","Std. Error"]  #extract the std error from the regression output
    P.Wald <- coef(summary(reg))["Dosage","Pr(>|z|)"]  #extract the p-value from the regression output
  }
  
# Report Sample size, the cases and controls for each variant as these values fluctuate across the study per person  
  Sample <- length(all$status_new)
  N.cases <- sum(all$status_new)
  N.controls <- Sample - N.cases
  
  MarkerName <- colnames(all[i])
  
# Consolidate all results in a single report for output
  Output <- cbind(MarkerName,N.cases,N.controls,Frequency,Estimate,SE,P.Wald,chi.stat.LRT,P.LRT,Score.Z,P.Score)
  write.table(Output, "BCAC_Overall_Status_Results.txt", append=T, quote=F, sep="\t", row.names=F, col.names=!file.exists("BCAC_Overall_Status_Results.txt"))    
}
```

### Results

Here's the output for the last variant that was processed / analyzed

```{r, warning = FALSE}
Output
```

Lets read in the complete results file next, with analyses results for all 3135 variants we analyzed

```{r, warning = FALSE}
results <- read.table("BCAC_Overall_Status_Results.txt", header=T, fill=T) 
results [1:5, 1:5] 
```

A little more post-processing. Before we can plot the results we have to merge in additional annotations for the variants so we can plot the results by chromosome. So read in the info file with the annotation data and merge it into the results file

```{r, warning = FALSE}
info <- read.table("BCAC_582_imputed_variants_info.txt", header=TRUE, fill=T) 
info [1:5, ] 

```

```{r, warning = FALSE}
tog <- merge(info, results, by.x="MarkerName", by.y="MarkerName", sort = TRUE)
tog[1:5,1:10] 
```

### Plots

Before we can create manhattan and q-q plots from the results, lets first examine how many variants are on each chromosome?
```{r, warning = FALSE}
as.data.frame(table(tog$chr))
```

```{r, warning = FALSE}
P = tog$P.Wald  #using the Wald p-value here as it is most commonly used
BP = tog$position
CHR = tog$chr

dat= cbind.data.frame(CHR,BP,P)
```

Manhattan plots represent the P values of the entire GWAS on a genomic scale. The P values are represented in genomic order by chromosome and position on the chromosome (x-axis). The value on the y-axis represents the -log10 of the P value (equivalent to the number of zeros after the decimal point plus one). Because of local correlation of the genetic variants, arising from infrequent genetic recombination, groups of significant P values tend to rise up high on the Manhattan plot, making the graph look like a Manhattan skyline.  We have lots of highly significant findings - most of which are among well known and previously reported genes & regions, but this is still a great postive test indicating our analyses, and our measurement values are all accurate and indicate sound findings. 

Note: To account for false positive rate, or the Type I error, genetic analyses commonly control this via a Bonferonni correction. The Bonferroni correction simply divides the significance level at each locus by the number of tests. In other words, it adjusts the alpha value from a = 0.05 to a =(0.05/k) where k is the number of statistical tests conducted. Most commonly used approximation: Significant associations have a P value smaller than 5E-8, under the assumption that 1 million independent tests are performed. The significance threshold 5E-8, also termed "genome-wide significance" is reached by dividing the usual alpha of 0.05 by 1 million (the effective number of tests performed). In our case, we have fewer variants than a complete GWAS (which would consists of multi-million variants), but we should still adhere to the 5E-8 as our variants were chosen from a set of high risk variant groups and more likely to be significantly associated with cancer risk.

```{r, warning = FALSE}
tiff(filename = "MH_plot.tiff", width= 8, height = 8, units="in", res=500)
manhattan(dat, col =c("blue", "gray60", "orange"), main = "Manhattan Plot of GWAS p-values", ylim = c(0, 200), cex = 0.6, 
    cex.axis = 0.9, genomewideline = -log10(5e-08), suggestiveline = F, chrlabs = c(1:23), annotateTop = TRUE)
dev.off()
```

The Quantile-Quantile plot is a graphical representation of the deviation of the observed P values from the null hypothesis: the observed P values for each variant are sorted from largest to smallest and plotted against expected values from a theoretical chi-square distribution.  It's a plot of the observed distribution of p-values vs. the expected distribution of p-values. We don't want a uniform distribution of p-values (indicated by a straight diagonal line), that would mean there are no causal variants. In our case, we see a sharp lift off due to the highly significant associations and nothing seems out of place.

```{r, warning = FALSE}
tiff(filename = "QQ_plot.tiff", width= 8, height = 8, units="in", res=500)
qq(P, main = "Q-Q plot of GWAS p-values", xlim = c(0, 7), ylim = c(0, 12), pch = 18, col = "blue4", cex = 1.5, las = 1)
dev.off()

```

The genomic inflation factor, lambda, is defined as the ratio of the median of the empirically observed distribution of the test statistic to the expected median, thus quantifying the extent of the bulk inflation and the excess false positive rate. The lambda values and Q-Q plots are used to compare the genome-wide distribution of the test statistic with the expected null distribution. The Q-Q plot is a useful visual tool to mark deviations of the observed distribution from the expected null distribution. Inflated lambda values or residual deviations in the Q-Q plot may point to undetected sample duplications, unknown familial relationships, a poorly calibrated test statistic, systematic technical bias or gross population stratification. 

Calculate the genomic inflation factor

```{r, warning = FALSE}
l_chisq <- qchisq (1-P, 1)

lambda <- median (l_chisq / qchisq(0.5,1))
lambda
```

A value of 1.83 is not a highly inflated value given our cherry-picked set of variants. So carry on-ward, all is well!

### Our findings

To take a closer look at our results, lets prune the complete set of analyes to the top hits that are genomewide
```{r}
top_hits <-subset(tog, P.Wald <5E-8)
write.table(top_hits, "BCAC_Overall_Status_Results_top_hits.txt", append=F, quote=F, sep="\t", row.names=F, col.names=T)

``` 

From our results, we see that well known genes such as FGFR2 (the most significantly associated region on chromosome 10) are among the well known variants we were able to replicate in our analyses. Other well known genes include BRCA2 (chr 13), FTO (chr 16), MAP3K1 (chr 5), MYC (chr 8), EMBP1 (chr 1) and IGFBP5 (chr 2).  Unfortunately we did not find any novel results that were globally significant. However, the sample of individuals recruited to the study were part of high risk population and having identified key genes that impact cancer risk among this sub-set can be used to create some form of a genetic score that can then be implemented in a clincal setting to flag/highlight like individuals who could be of a higher breast cancer risk profile.
 